server/src/main/java/org/apache/lucene/index/LazySoftDeletesDirectoryReaderWrapper.java

```diff
@@ -0,0 +1,288 @@
+        final SegmentCommitInfo segmentInfo = segmentReader.getSegmentInfo();
+        final int numSoftDeletes = segmentInfo.getSoftDelCount();
+        if (numSoftDeletes == 0) {
+            return reader;
+        }
+        final int maxDoc = reader.maxDoc();
+        final int numDocs = maxDoc - segmentInfo.getDelCount() - segmentInfo.getSoftDelCount();
+        final LazyBits lazyBits = new LazyBits(maxDoc, field, reader, numSoftDeletes, numDocs);
+        return reader instanceof CodecReader ? new LazySoftDeletesFilterCodecReader((CodecReader) reader, lazyBits, numDocs)
+            : new LazySoftDeletesFilterLeafReader(reader, lazyBits, numDocs);
+    }
+
+    public static final class LazySoftDeletesFilterLeafReader extends FilterLeafReader {
+        private final LeafReader reader;
+        private final LazyBits bits;
+        private final int numDocs;
+        private final CacheHelper readerCacheHelper;
+
+        public LazySoftDeletesFilterLeafReader(LeafReader reader, LazyBits bits, int numDocs) {
+            super(reader);
+            this.reader = reader;
+            this.bits = bits;
+            this.numDocs = numDocs;
+            this.readerCacheHelper = reader.getReaderCacheHelper() == null ? null :
+                new DelegatingCacheHelper(reader.getReaderCacheHelper());
+        }
+
+        @Override
+        public LazyBits getLiveDocs() {
+            return bits;
+        }
+
+        @Override
+        public int numDocs() {
+            return numDocs;
+        }
+
+        @Override
+        public CacheHelper getCoreCacheHelper() {
+            return reader.getCoreCacheHelper();
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            return readerCacheHelper;
+        }
+    }
+
+    public static final class LazySoftDeletesFilterCodecReader extends FilterCodecReader {
+        private final LeafReader reader;
+        private final LazyBits bits;
+        private final int numDocs;
+        private final CacheHelper readerCacheHelper;
+
+        public LazySoftDeletesFilterCodecReader(CodecReader reader, LazyBits bits, int numDocs) {
+            super(reader);
+            this.reader = reader;
+            this.bits = bits;
+            this.numDocs = numDocs;
+            this.readerCacheHelper = reader.getReaderCacheHelper() == null ? null :
+                new DelegatingCacheHelper(reader.getReaderCacheHelper());
+        }
+
+        @Override
+        public LazyBits getLiveDocs() {
+            return bits;
+        }
+
+        @Override
+        public int numDocs() {
+            return numDocs;
+        }
+
+        @Override
+        public CacheHelper getCoreCacheHelper() {
+            return reader.getCoreCacheHelper();
+        }
+
+        @Override
+        public CacheHelper getReaderCacheHelper() {
+            return readerCacheHelper;
+        }
+    }
+
+    private static class DelegatingCacheHelper implements CacheHelper {
```

Comment:
I believe that this bit is the only reason why we need to have this class in the `oal.index` package. Can we find a way to avoid doing cross-JAR package-protected access? (Is there another option than making the CacheKey ctor public?)

Reply:
The only other reason except for `CacheKey` is the call to the `PendingSoftDeletes.applySoftDeletes` method. That method can easily be reimplemented, however. I don't see a good way to work around the `CacheKey` constructor not being accessible (safe for introducing even more outrageous hacks), and FWIW we already have some classes like `ShuffleForcedMergePolicy` and `OneMergeHelper` in this package.